src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -30,21 +32,432 @@
+            Assert.Same(newSolution1, newSolution2);
+
+            Assert.Throws<ArgumentNullException>(() => solution.WithDocumentText(documentId, (SourceText)null!, PreservationMode.PreserveIdentity));
+            Assert.Throws<ArgumentOutOfRangeException>(() => solution.WithDocumentText(documentId, text, (PreservationMode)(-1)));
+
+            Assert.Throws<ArgumentNullException>(() => solution.WithDocumentText((DocumentId)null!, text, PreservationMode.PreserveIdentity));
+            Assert.Throws<InvalidOperationException>(() => solution.WithDocumentText(s_unrelatedDocumentId, text, PreservationMode.PreserveIdentity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        public void WithDocumentText_TextAndVersion()
+        {
+            var solution = CreateSolutionWithProjectAndDocuments();
+            var documentId = solution.Projects.Single().DocumentIds.Single();
+            var textAndVersion = TextAndVersion.Create(SourceText.From("new text"), VersionStamp.Default);
+
+            var newSolution1 = solution.WithDocumentText(documentId, textAndVersion, PreservationMode.PreserveIdentity);
+            Assert.True(newSolution1.GetDocument(documentId)!.TryGetText(out var actualText));
+            Assert.True(newSolution1.GetDocument(documentId)!.TryGetTextVersion(out var actualVersion));
+            Assert.Same(textAndVersion.Text, actualText);
+            Assert.Equal(textAndVersion.Version, actualVersion);
+
+            var newSolution2 = newSolution1.WithDocumentText(documentId, textAndVersion, PreservationMode.PreserveIdentity);
+            Assert.Same(newSolution1, newSolution2);
+
+            Assert.Throws<ArgumentNullException>(() => solution.WithDocumentText(documentId, (SourceText)null!, PreservationMode.PreserveIdentity));
+            Assert.Throws<ArgumentOutOfRangeException>(() => solution.WithDocumentText(documentId, textAndVersion, (PreservationMode)(-1)));
+
+            Assert.Throws<ArgumentNullException>(() => solution.WithDocumentText((DocumentId)null!, textAndVersion, PreservationMode.PreserveIdentity));
+            Assert.Throws<InvalidOperationException>(() => solution.WithDocumentText(s_unrelatedDocumentId, textAndVersion, PreservationMode.PreserveIdentity));
         }
 
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        public void WithDocumentText_MultipleDocuments()
+        {
+            var solution = CreateSolutionWithProjectAndDocuments();
+            var documentId = solution.Projects.Single().DocumentIds.Single();
+            var text = SourceText.From("new text");
+
+            var newSolution1 = solution.WithDocumentText(new[] { documentId }, text, PreservationMode.PreserveIdentity);
+            Assert.True(newSolution1.GetDocument(documentId)!.TryGetText(out var actualText));
+            Assert.Equal(text, actualText);
```

Comment:
```suggestion
            Assert.Same(text, actualText);
```

Since you're using PreserveIdentity?